Durable AI Agent Tutorial (#399)
temporalio/temporal-learning: docs/tutorials/ai/durable-ai-agent/index.md

:::

* Text editor or IDE of your choice to write code.

### Optional

You can opt to use real API services for your tools, or use provided mock functions that will also be provided. 

* A free [RapidAPI Sky Scrapper API Key](https://rapidapi.com/apiheya/api/sky-scrapper) saved securely where you can access it. You will use this to search for flights.
* A free [Stripe Account](https://stripe.com/lp/start-now) with a configured [sandbox](https://docs.stripe.com/sandboxes). You will use this to generate fake invoices for the flights that are being booked.

### Concepts

Additionally, this tutorial assumes you have basic familiarity with:

#### Programming Concepts

* Temporal fundamentals such as [Workflows](https://docs.temporal.io/develop/python/core-application#develop-workflows), [Activities](https://docs.temporal.io/develop/python/core-application#develop-activities), [Workers](https://docs.temporal.io/develop/python/core-application#run-a-dev-worker), [Signals](https://docs.temporal.io/develop/python/message-passing#signals), and [Queries](https://docs.temporal.io/develop/python/message-passing#queries)

> * Python fundamentals such as functions, classes, async/await syntax, and virtual environments

Reviewer: I think this would potentially be out of scope, since we'd be linking outside of Temporal's body of knowledge and we'd have to arbitrarily pick winners elsewhere on the internet for a lot of them.

> * Python fundamentals such as functions, classes, async/await syntax, and virtual environments

Author: I think with this tutorial we have to assume the reader is much further advanced than our typical expectation. This tutorial assumes prior understanding of Temporal and AI Agents. I will add a link to asyncio docs for the Python docs, as that's the most likely things for Python devs to be unfamiliar with, but the rest is almost table stakes.

Now that you've defined the `ToolDefinition` that will be available for your agent, you can define the `AgentGoal` type and create your agent's goal.

### Definining the `AgentGoal` type

To define the `AgentGoal` type, open `models/core.py` and add the following code:

```python
@dataclass
class AgentGoal:
    agent_name: str
    tools: List[ToolDefinition]
    description: str
    starter_prompt: str
    example_conversation_history: str
```

This `dataclass` defines your `AgentGoal` as a combination of a few attributes:
* `agent_name` - A human readable name for the agent
* `tools` - A list of tools, defined as `ToolDefinition` types, that the agent can use to achieve its goal
* `description` - A description of the goal, in a bulleted list format specifying how to achieve it.
* `starter_prompt` - A starter prompt for the AI agent to run
* `example_conversation_history` - A sample conversation history of what a successful interaction with this agent would look like

<details>

<summary>
The <code>models/core.py</code> is complete and will need no more revisions. You can review the complete file and copy the code here
</summary>

<br />

> Hover your cursor over the code block to reveal the copy-code option.

Reviewer 2: This is the first time we've included this instruction. Think we should include it on every one of these code boxes, or just on the first one?

```python
from dataclasses import dataclass
from typing import List

@dataclass
class ToolArgument:
    name: str
    type: str
    description: str

@dataclass
class ToolDefinition:
    name: str
    description: str
    arguments: List[ToolArgument]

@dataclass
class AgentGoal:
    id: str
    category_tag: str
    agent_name: str
    agent_friendly_description: str
    tools: List[ToolDefinition]
    description: str = "Description of the tools purpose and overall goal"
    starter_prompt: str = "Initial prompt to start the conversation"
    example_conversation_history: str = (
        "Example conversation history to help the AI agent understand the context of the conversation"
    )
```
</details>

Now that you have the type available to define the goal, you will implement the goal for your agent.

### Implementing the goal registry

Similar to implementing the `tool_registry`, next you will implement a `goal_registry` to define your agent's goal and make it available to the Workflow. 
You will do this by creating an instance of your `AgentGoal` type for every goal you wish to implement.

> For this tutorial you will only implement a single goal, but you may want to use this framework going forward to create your own agent goals at a later date.

Reviewer 2: I suggest mentioning specifically here that the goal is called `goal_event_flight_invoice`. I admit that I was initially confused by the name. "Event" is kind of an overloaded term, and I thought you were referring to a programmatic event that occurs when an invoice is generated for a specific flight.

Author: Will do.

To implement your agent's goal, create the file `models/goal_registry.py` and add the following imports to the file:

```python
import tools.tool_registry as tool_registry
from models.core import AgentGoal
```

To create the goal, first create an instance of the `AgentGoal` `dataclass` and add the first parameter, `agent_name`, to identify the goal:

```python
goal_event_flight_invoice = AgentGoal(
    agent_name="North America Event Flight Booking",
    # ...
```

> Next, pass in the tools that the agent is allowed to use to accomplish its goal to the `tools` parameter.

Reviewer 2: Is it worth mentioning that you're specifically passing references to the tool registry, as opposed to the files for the tools or anything else?

Add it as the next parameter when creating the `goal_event_flight_invoice` object.

```python
    # ...
    tools=[
        tool_registry.find_events_tool,
        tool_registry.search_flights_tool,
        tool_registry.create_invoice_tool,
    ],
    # ...
```

The following parameter defines a detailed description of what the goal is and the ideal path for the agent to take to achieve its goal.
Add it as the next parameter when creating the `goal_event_flight_invoice` object.

```python
    # ...
    description="Help the user gather args for these tools in order: "
    "1. FindEvents: Find an event to travel to "
    "2. SearchFlights: search for a flight around the event dates "
    "3. CreateInvoice: Create a simple invoice for the cost of that flight ",
    # ...
```

> The next parameter provides a starter prompt for the agent, detailing how it should begin its interaction with every user.

Reviewer 2: The starter prompt and conversation parameters threw me off a bit. They seem very free-form, and not programmatic. It feels like there could be some guidance, or a link to some guidance, as to how to create these elements. The sample conversation especially. Is there standard syntax for that?

Author: Ya, there's no standard on this. It seems to just be "vibes". I will add in a brief explanation of what a starter prompt is and how you should craft one.

Add it as the next parameter when creating the `goal_event_flight_invoice` object.

```python
    # ...
    starter_prompt="Welcome me, give me a description of what you can do, then ask me for the details you need to do your job.",
    # ...
```

Finally, draft an example conversation of a successful interaction with your agent to pass in.
LLMs perform better when they have an example of expected output, so providing this aids the LLM in its goal.
Since this is a `str` type, but the conversation is long, you will define each statement as a line in a list and then use `"\n ".join()` to create a string from your conversation.
Add the conversation as the final parameter when creating the `goal_event_flight_invoice` object.

```python
    # ...
    example_conversation_history="\n ".join(
        [
            "user: I'd like to travel to an event",
            "agent: Sure! Let's start by finding an event you'd like to attend. I know about events in North American cities. Could you tell me which city and month you're interested in?",

>             "user: sydney in may please",

Reviewer 2: Is this an error, or did you intend for the user to ask for Sydney (outside North America)?

Author: Error. Original version was in Australia. I brought it home

            "agent: Great! Let's find an events in New York City in May.",
            "user_confirmed_tool_run: <user clicks confirm on FindEvents tool>",

>             "tool_result: { 'event_name': 'Vivid New York City', 'event_date': '2023-05-01' }",

Reviewer 2: Now I'm just nitpicking, but "Vivid Sydney" is a Sydney-specific event. The only search result for "Vivid New York City" is a cabaret that I wouldn't mention in an all-ages tutorial.

├── pyproject.toml
├── uv.lock
├── models/
│   ├── __init__.py
│   └── core.py
├── scripts/
│   ├── create_invoice_test.py
│   ├── find_events_test.py
│   └── search_flights_test.py
└── tools/
    ├── __init__.py
    ├── create_invoice.py
    ├── find_events.py
    ├── goal_registry.py
    ├── search_flights.py
    ├── tool_registry.py
    └── data/
        └── find_events_data.json
```
</details>

## Building Temporal Activities to execute non-deterministic agent code

Now that you have built the agent's goal, the tools it needs to achieve it, you can start building the agent code. 
In this step, you will create Activities that execute code in your AI agent that can behave non-deterministically, such as making the LLM calls or calling tools..
As tools can call out to external services, have the possibility to fail, be rate limited, or perform other non-deterministic operations, it's safer to always call them in an Activity.
When an Activity fails, they're by default automatically retried until it succeeds or is canceled.
Another added benefit of executing your tool as an Activity is after the Activity completes, the result is saved to an Event History managed by Temporal.

> If your application were to then crash after executing a few tools, it could reconstruct the state of the execution without having to re-execute the tools and use the previous executions results.

Author: Works for me.

This provides durability to your agent for intermittent issues, which are common in distributed systems.

> Before you can proceed to creating the Activities, however, you need to create the custom types that you'll use for Activity communication.

Author: Activities having one parameter is how we can provide type safety. I'll add something here to reiterate that.

Next, add the `dataclass` that handles the input for calling the LLM for tool planning:

```python
@dataclass
class ToolPromptInput:
    prompt: str
    context_instructions: str
```

`ToolPromptInput` contains the prompt the Activity will issue to the LLM, along with any context that the LLM needs when executing the prompt.

To go along with the this type, you need to add types that store the results of validation of the prompt:

```python
@dataclass
class ValidationInput:
    prompt: str
    conversation_history: ConversationHistory
    agent_goal: AgentGoal

@dataclass
class ValidationResult:
    validationResult: bool
    validationFailedReason: Dict[str, Any] = field(default_factory=dict)
```

The `ValidationInput` type contains the prompt given by the user, the conversation history, and the agent's goal.
An Activity will use this type as input and validate the prompt against the agent's goal.
Conversely, the `ValidationResult` type will contain the results of this validation Activity and will return a boolean signifying if the prompt passed or failed, and if it did fail a reason why.

> Next, add two more `dataclass`es for handling the input and output of reading environment variables into the Workflow:

Author: Added

        return response_content
```

The second helper function takes a string as input and returns a dictionary after attempting to parse the string as valid JSON.
Add this method to the bottom of your `activities.py` file:

```python
    def parse_json_response(self, response_content: str) -> dict:
        """
        Parses the JSON response content and returns it as a dictionary.
        """
        try:
            data = json.loads(response_content)
            return data
        except json.JSONDecodeError as e:
            activity.logger.error(f"Invalid JSON: {e}")
            raise
```

Now that you have the helper methods implemented, you can implement the Activity responsible for making LLM calls.

#### Implementing the Activity for making LLM calls

The `agent_toolPlanner` Activity handles all interactions with your chosen LLM.
It makes the call to the LLM, parses the response and returns JSON on success, and raises an Exception on failure.

Add the method header with the appropriate decorator to your `activities.py` file, underneath the `__init__` method:

```python
    @activity.defn
    async def agent_toolPlanner(self, input: ToolPromptInput) -> dict:
```

> Next, create the `messages` list, which contains various dictionaries to the specification of the LLM for prompting.

Author: Added a link to the LiteLLM docs that show how to do this. I also updated that phrase to be less stuff.

```command
cd frontend
npx vite
```

You will see output to your terminal, and then your web browser will open to `localhost:5173` with your agent running.

:::note

When first starting the web UI, you may see a red error banner appear upon startup with a message about timeouts.
This is expected, as the UI begins polling immediately before the Workflow may begin.
This will go away within a few seconds once the Workflow Execution has started and the first message from the agent appears.

:::

Finally, open a new browser tab and navigate to `localhost:8233`.
This will display the Temporal Web UI.
You should see a running Workflow Execution there with the Workflow ID **agent-workflow**.
Click on the link to open it so you can watch the Workflow progress as you run test your agent.

### Testing the complete system

With all components running, you can now test the agent Workflow.

Navigate back to `localhost:5173`.
You should see a message _similar_ to the following.
Remember, the agent's responses are powered by an LLM, so the responses are non-deterministic, meaning they are likely to be slightly different every time.

```text

> Agent: Welcome! I'm here to help you plan your travel to events in North America. I can assist you in finding events, booking flights, and generating invoices for your trip. To get started, please tell me which city and month you're interested in traveling to?

Author: Sure

## Setting up your development environment

Before you start coding, you need to set up your Python developer environment.
In this step, you will set up your project structure, install the necessary Python packages, and configure the Python environment needed to build your AI agent.

First, create your project using `uv`:

```command
uv init temporal-ai-agent --python ">=3.9"

```

Next, change directories into your newly created project:

```command
cd temporal-ai-agent
```

`uv` is a modern Python project and packaging tool that sets up a project structure for  you.
Running this command creates the following default Python package structure for you:

```
temporal-ai-agent/
├── .gitignore
├── .python-version
├── README.md
├── pyproject.toml
└── uv.lock
```

It automatically runs a `git init` command for you, provides you with the default `.gitignore` for Python, creates a `.python-version` file that has the project's default Python version, a README.md, a Hello World `main.py` program, and a `pyproject.toml` file for managing the projects packages and environment.

> You won't need the `main.py` file, so delete it:

Reviewer: This wasn't listed in the above tree which is a bit confusing

Author: Good catch. Will add in.

# Temporal Configuration
TEMPORAL_ADDRESS=localhost:7233
TEMPORAL_NAMESPACE=default
TEMPORAL_TASK_QUEUE=agent-task-queue

# (Optional) - Uncomment both lines and set RAPIDAPI_KEY if you plan on 
# using the real flights API
# RAPIDAPI_KEY=YOUR_RAPID_API_KEY
# RAPIDAPI_HOST_FLIGHTS=sky-scrapper.p.rapidapi.com

# (Optional) - Uncomment and set STRIPE_API_KEY if you plan on using the Stripe
# API to generate a fake invoice
# STRIPE_API_KEY=YOUR_STRIPE_API_KEY

# Uncomment if connecting to Temporal Cloud using mTLS (not needed for local dev server)
# TEMPORAL_TLS_CERT='path/to/cert.pem'
# TEMPORAL_TLS_KEY='path/to/key.pem'

# Uncomment if connecting to Temporal Cloud using API key (not needed for local dev server)
# TEMPORAL_API_KEY=abcdef1234567890
```

Once copied, replace `YOUR_OPEN_API_KEY` with your OpenAI API key.
Setting `SHOW_CONFIRM=True` requires the user to confirm each tool prior to it being executed.
This will allow you to see what the agent is doing step by step.
These are the only two mandatory variables to set.
This tutorial provides both an ability to create pseudo tools that perform simulations, or tools that use external APIs to achieve their goals.

> If you plan on using the [RapidAPI SkyScraper API](#optional) to look up flight data or the [Stripe API](#optional) to generate an invoice, you can uncomment these lines and provide the API keys here.

Reviewer: Should this be an if? Aren't these the main two APIs we're demoing with this tutorial? I think Temporal Cloud in the next paragraph should be an if, this sentence should just start with "To use the"

Author: The tutorial defaults to having these not set. The only required API key is an LLM key. After considering it, the value of this tutorial is the tool execution orchestration with the LLMs, so that's the primary focus. By providing an API key, it will use the "live" version, but if you don't, it will use a stubbed out version. This keeps people from needing to have 3 different accounts to run things, and safeguards against the flakiness I've dealt with regarding the Flights API. 

Long way of saying "Yes, it should be if here."

> 

Reviewer: This whitespace doesn't seem to be rendered which makes the preceding paragraph too long... maybe just throw a `<br/>` in here, idk what Docusaurus is doing.

### Setting up the `tools` package

To get started, first create the directory for your tools modules:

```command
mkdir tools
```

Then change directories into it:

```command
cd tools
```

However, for this to be an importable tools package, you will need to add a `__init__.py` file.
It can be blank for now, so create it with the following command:

```command
touch __init__.py
```

Now that you have set up the structure for your tools package, you'll acquire and test the tools needed to have the agent succeed with its goal.

### Acquiring the `find_events` tool

The `find_events` tool searches for events within a given city during a certain time of year.
The tool takes a month and city as inputs and provides events for not only the month that was provided, but the months before and after the given month as well. 
The LLM will use this tool to search for events when helping the user plan their trip.
This tool doesn't use an API, but rather simulates looking events up in a data store using mock data.

First, create a `data` directory within the `tools` directory to store the sample event data and change directories into it:

```command
mkdir data
cd data
```

> Next, run the following command to download the sample data from the [companion GitHub repository](https://github.com/temporal-community/tutorial-temporal-ai-agent):

Reviewer: I notice this is unmerged, did you want a review on this too?

You will also need to uncomment the `RAPIDAPI_HOST_FLIGHTS` environment variable as this is the endpoint the tool will be accessing.

```bash
RAPIDAPI_KEY=YOUR_RAPID_API_KEY
RAPIDAPI_HOST_FLIGHTS=sky-scrapper.p.rapidapi.com
```

Next, review the code in `scripts/search_flights_test.py` and make sure that the `dateDepart` and `dateReturn` dates are both in the future.
At this point you have no way of determining if the dates are in the past, and the API will return an error if you try to search for flights in the past. 

Once you've reviewed the code, run the test using the following command:

```command
uv run scripts/search_flight_test.py
```

If you've changed the dates or cities, you may see different results, but the format should be similar to this:

```output
Searching for: ORD
Searching for: DFW
{
  "origin": "ORD",
  "destination": "DFW",
  "currency": "USD",
  "results": [
    {
      "outbound_flight_code": "NK824",
      "operating_carrier": "Spirit Airlines",
      "return_flight_code": "NK828",
      "return_operating_carrier": "Spirit Airlines",
      "price": 119.98
    },
  ]
}
```

> :::warning

Reviewer: This renders as "DANGER" which is I think a little too forceful here; this can be an `info` block instead I think.

    print(invoice_details)
```

However, since this tool can operate in either a mock mode or live API mode, there are two ways to verify it.

#### Testing the mocked `create_invoice` tool

Start by testing it without the Stripe key. 
If you have it set in your `.env` file, comment it out for now, or skip this step.

Run the test using the following command:

```command
uv run scripts/search_flight_test.py
```

The output should be:

```output
[CreateInvoice] Creating invoice with: {'email': 'ziggy.tardigrade@example.com', 'amount': 150.0, 'description': 'Flight to Replay', 'days_until_due': 7}
{'invoiceStatus': 'generated', 'invoiceURL': 'https://pay.example.com/invoice/12345', 'reference': 'INV-12345'}
```

If you aren't planning on using the Stripe API, you can skip this next step and continue if you'd like.

#### Testing the Stripe-powered `create_invoice` tool

Testing the Stripe powered version of the tool is nearly identical to testing the mocked version of the tool.

First, if you haven't uncommented the `STRIPE_API_KEY` lines in your `.env` file and added your API key, do this before running the test.

```bash
STRIPE_API_KEY=YOUR_STRIPE_API_KEY
```

:::warning

> Make sure you have set up your Stripe account as a [sandbox](https://docs.stripe.com/sandboxes) and are using an API key from there.

Reviewer: ```suggestion
Make sure you have set up your Stripe account as a [sandbox](https://docs.stripe.com/sandboxes) and are using an API key from there. If it is your first time setting up a Stripe account and you haven't added any billing information, this will be the default.
```

"""
```

This segment provides strict rules on the exact format the LLM should respond with, as well as guardrails to ensure that fields are set properly.
The guardrails section is particularly important as it provides detailed behavioral constraints that enable consistent responses. 
These rules prevent issues such as asking questions while proposing tool execution or forgetting to use the conversation history for argument inference.

Finally, complete the template with a validation prompt:

```python
"""
{% if raw_json is not none %}

=== Validation Task ===
Validate and correct the following JSON if needed:
{{ raw_json_str }}

Check syntax, 'tool' validity, 'args' completeness, and set 'next' appropriately. Return ONLY corrected JSON.
{% endif %}

{% if raw_json is not none %}
Begin by validating the provided JSON if necessary.
{% else %}
Begin by producing a valid JSON response for the next tool or question.
{% endif %}
""".strip()
)
```

The validation section enables the template to handle both correct and incorrectly JSON formatted strings.
If the JSON is improperly formatted, the LLM is prompted to correct it before continuing with its other tasks.

> All together, the template should look as such:

Reviewer: ```suggestion
All together, the template should look like this:
```

    Returns:
        str: A formatted prompt string for requesting missing arguments
    """
    return MISSING_ARGS_PROMPT.format(
        response=tool_data.get("response"),
        current_tool=current_tool,
        missing_args=missing_args,
    )
```
</details>

<details>

<summary>
Before moving on to the next section, verify your files and directory structure is correct.
</summary>

```
temporal-ai-agent/
├── .env.example
├── .gitignore
├── .python-version
├── README.md
├── pyproject.toml
├── uv.lock
├── activities/
|   ├── __init__.py
|   └── activities.py
├── models/
│   ├── __init__.py
│   ├── core.py
│   └── requests.py
├── prompts/
│   ├── __init__.py
│   ├── agent_prompt_generators.py
│   └── prompts.py
├── scripts/
│   ├── create_invoice_test.py
│   ├── find_events_test.py
│   └── search_flights_test.py
└── tools/
    ├── __init__.py
    ├── create_invoice.py
    ├── find_events.py
    ├── goal_registry.py
    ├── search_flights.py
    ├── tool_registry.py
    └── data/
        └── find_events_data.json
```
</details>

Now that you have the prompt rendering submodule implemented, you can implement the main agent Workflow.

> ## Building the agent Workflow

Reviewer: General comment -- this tutorial is *really long* and repetitive in places and this is the first thing approaching a real breakpoint, but it's not a great breakpoint because there's no end-to-end output that can be produced up to now. Can we do anything about that? I think we have to get this off of a single page somehow, I know our best examples of that (eg the background check in Go example) haven't been very successful at it either, but it's hard to imagine teaching this end-to-end right now, let alone someone dutifully following the whole thing on a single page.

            start_to_close_timeout=TOOL_ACTIVITY_START_TO_CLOSE_TIMEOUT,
            retry_policy=RetryPolicy(
                initial_interval=timedelta(seconds=5), backoff_coefficient=1
            ),
        )
        dynamic_result["tool"] = current_tool
    except ActivityError as e:
        workflow.logger.error(f"Tool execution failed: {str(e)}")
        dynamic_result = {"error": str(e), "tool": current_tool}

    add_message_callback("tool_result", dynamic_result)
    prompt_queue.append(generate_tool_completion_prompt(current_tool, dynamic_result))
```
It executes the tool by calling the name of the tool, which gets handled by the dynamic Activity you implemented.
When calling the Activity, you specify the Activity timeouts using the constants you defined earlier.
Whether the Activity succeeds or fails, the result is stored to the conversation history using the `add_message_callback` that was passed in.
Then, the method invokes the `generate_tool_completion_prompt` function with the `current_tool` and result of the tool execution to create a prompt and add it to the `prompt_queue`, which the agent will handle on its next iteration.

> #### Defining the missing argument handler function

Author: I do wish I had those. This is why I have the "Here's the finished block of code" once the file no longer needs edits. It will have to suffice until we have the time to implement the Markdown components we need.

If so, the agent will return the conversation history stored in the `self.conversation` instance variable, and the Workflow Execution will close.

#### Executing the tool

Next, your agent will determine if it is appropriate to execute a the tool, and if it is, invoke an Activity to do so.

Continue by adding the following code to execute the tool:

```python
            # Execute the tool
            if self.ready_for_tool_execution() and current_tool is not None:
                await self.execute_tool(current_tool)
                continue
```

Before the agent executes a tool, the agent confirms that the tool meets the requirements for execution and that the current tool is not `None`.
If both of these checks evaluate to `True`, the agent executes the tool.
Once the tool has completed execution, it `continue`s the loop, meaning it skips all further execution and returns to the top of the loop, ready to begin another iteration.

#### Adding in a few more helper methods

Next, implement three helper method that the tool execution code block called, but had not yet implemented.

The first checks if the tool is ready for execution.

> Add this method within your class but outside of the `run` method, underneath any other previously defined methods.

Author: Ya, this isn't fun...I could probably say append the method within the class. I have a tutorial like this from the DO days that does this and it did ok. Hopefully the giant "here's the finished code" alleviates this. 

I think I can change it to "Add this method to your class".  I think the "Outside of the run method" is unnecessary, because I've never really seen anyone implement a method within another method. And the underneath any other previously defined methods seems excessive.

It is waiting for Workflows and Activity tasks to execute.
If your Worker is running successfully, that's as much as you can test for the moment.
Kill both the worker and Temporal service by pressing `CTRL-C` in each terminal.

<details>

<summary>
Before moving on to the next section, verify that your files and directory structure are correct.
</summary>

```
temporal-ai-agent/
├── .env.example
├── .gitignore
├── .python-version
├── README.md
├── pyproject.toml
├── uv.lock
├── activities/
|   ├── __init__.py
|   └── activities.py
├── models/
│   ├── __init__.py
│   ├── core.py
│   └── requests.py
├── prompts/
│   ├── __init__.py
│   ├── agent_prompt_generators.py
│   └── prompts.py
├── scripts/
│   ├── create_invoice_test.py
│   ├── find_events_test.py
│   └── search_flights_test.py
├── tools/
│   ├── __init__.py
│   ├── create_invoice.py
│   ├── find_events.py
│   ├── goal_registry.py
│   ├── search_flights.py
│   ├── tool_registry.py
│   └── data/
|       └── find_events_data.json
├── worker/
│   └── worker.py
└── workflows/
    ├── __init__.py
    ├── agent_goal_workflow.py
    └── workflow_helpers.py
```

</details>

> Next, you will implement a REST API that will serve as the backend service for invoking your agent.

Reviewer: Even though this is like 75% of the way through I'd strongly consider stopping here with an example of how to interact with this thing purely using the command line and putting the UI in a followup tutorial. It's one of the only logical breakpoints we have.

│   ├── core.py
│   └── requests.py
├── prompts/
│   ├── __init__.py
│   ├── agent_prompt_generators.py
│   └── prompts.py
├── scripts/
│   ├── create_invoice_test.py
│   ├── find_events_test.py
│   └── search_flights_test.py
├── tools/
│   ├── __init__.py
│   ├── create_invoice.py
│   ├── find_events.py
│   ├── goal_registry.py
│   ├── search_flights.py
│   ├── tool_registry.py
│   └── data/
|       └── find_events_data.json
├── worker/
│   └── worker.py
└── workflows/
    ├── __init__.py
    ├── agent_goal_workflow.py
    └── workflow_helpers.py
```

</details>

In the next step, you will test your agent using a chatbot web interface.

## Running your agent 

Now that you have implemented a mechanism of communication for your agent, it's time to test it.
You will now download a React frontend that implements a chatbot UI to interact with your agent.
The UI will open in a terminal window and prompt the user with a message stating their purpose and instructing the user what to do next.
Throughout the conversation, the user will interact with the agent, responding to questions from the agent as the agent tries to accomplish its goal.

### Adding a Chatbot Web UI

> To get started, download the pre-built React based web UI:

Author: I did not. Thankfully this user will only run this for a few minutes on their machine, not hours at a booth, so it shouldn't kill the battery as fast. I did however decrease the polling interval to not be quite so obnoxious.
